File: backend/models/DonationAppointment.js
// backend/models/DonationAppointment.js
import mongoose from "mongoose";

const donationAppointmentSchema = new mongoose.Schema(
  {
    donor: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "Donor",
      required: true,
    },
    camp: {
      type: mongoose.Schema.Types.ObjectId,
      ref: "BloodCamp",
      required: true,
    },
    appointmentDate: {
      type: Date,
      required: true,
    },
    appointmentTime: {
      type: String,
      required: true,
    },
    status: {
      type: String,
      enum: ["pending", "confirmed", "checked_in", "completed", "cancelled"],
      default: "pending",
    },
    qrCode: {
      type: String,
    },
    checkInTime: Date,
    notes: String,
    cancellationReason: String,
    // backend/models/DonationAppointment.js
// Thêm vào schema
queueNumber: { type: Number, default: null },
bedNumber: { type: String, default: null },  // Số giường/bàn hiến máu
calledAt: { type: Date, default: null },     // Thời gian gọi donor
calledStatus: { 
  type: String, 
  enum: ['pending', 'called', 'in_progress', 'completed'],
  default: 'pending' 
},
completedAt: { type: Date, default: null },
completedBy: { type: mongoose.Schema.Types.ObjectId, ref: 'User' },
  },
  { timestamps: true }
);

export default mongoose.model("DonationAppointment", donationAppointmentSchema);

File: backend-go/internal/models/donation_appointment.go
package models

import (
	"go.mongodb.org/mongo-driver/bson/primitive"
	"time"
)

type DonationAppointment struct {
	ID                 primitive.ObjectID  `bson:"_id,omitempty" json:"id,omitempty"`
	Donor              primitive.ObjectID  `bson:"donor" json:"donor"`
	Camp               primitive.ObjectID  `bson:"camp" json:"camp"`
	AppointmentDate    time.Time           `bson:"appointmentDate" json:"appointmentDate"`
	AppointmentTime    string              `bson:"appointmentTime" json:"appointmentTime"`
	Status             string              `bson:"status" json:"status"` // pending, confirmed, checked_in, completed, cancelled
	QRCode             string              `bson:"qrCode,omitempty" json:"qrCode,omitempty"`
	CheckInTime        *time.Time          `bson:"checkInTime,omitempty" json:"checkInTime,omitempty"`
	Notes              string              `bson:"notes,omitempty" json:"notes,omitempty"`
	CancellationReason string              `bson:"cancellationReason,omitempty" json:"cancellationReason,omitempty"`
	QueueNumber        *int                `bson:"queueNumber,omitempty" json:"queueNumber,omitempty"`
	BedNumber          string              `bson:"bedNumber,omitempty" json:"bedNumber,omitempty"`
	CalledAt           *time.Time          `bson:"calledAt,omitempty" json:"calledAt,omitempty"`
	CalledStatus       string              `bson:"calledStatus" json:"calledStatus"` // pending, called, in_progress, completed
	CompletedAt        *time.Time          `bson:"completedAt,omitempty" json:"completedAt,omitempty"`
	CompletedBy        *primitive.ObjectID `bson:"completedBy,omitempty" json:"completedBy,omitempty"`
	CreatedAt          time.Time           `bson:"createdAt" json:"createdAt"`
	UpdatedAt          time.Time           `bson:"updatedAt" json:"updatedAt"`
}
